File: go/merge_sort.go
package main

import "fmt"

func main() {
	list := []int{5, 1, 4, 6, 3, 2}
	fmt.Println(list)
	sortedList := mergeSort(list)
	fmt.Println(sortedList)
}

func mergeSort(list []int) []int {
	if length := len(list); length > 1 {
		middle := length / 2
		return merge(mergeSort(list[0:middle]), mergeSort(list[middle:length]))
	}

	return list
}

func merge(left []int, right []int) []int {
	sortedList := make([]int, 0)

	for len(left) > 0 && len(right) > 0 {
		if left[0] > right[0] {
			sortedList = append(sortedList, right[0])
			right = right[1:]
		} else {
			sortedList = append(sortedList, left[0])
			left = left[1:]
		}
	}

	return append(sortedList, append(left, right...)...)
}

File: javascript/merge_sort.js
function merge_sort(arr) {
    if (arr.length > 1) {
        const middle = arr.length / 2;

        const left = merge_sort(arr.slice(0, middle))
        const right = merge_sort(arr.slice(middle))

        return merge(left, right)
    }

    return arr;
}

function merge(left, right) {
    const sorted = [];

    while (left.length && right.length) {
        if (left[0] < right[0]) {
            sorted.push(left.shift())
        } else {
            sorted.push(right.shift())
        }
    }

    return sorted.concat([...left,...right])
}

const array = [2,1,6,4,5,3];

console.log("Original", array);
console.log("Sorted", merge_sort([...array]));
